core/src/main/java/org/elasticsearch/cluster/routing/UnassignedMeta.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.routing;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.joda.FormatDateTimeFormatter;
+import org.elasticsearch.common.joda.Joda;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Holds additional information as to why the shard is in unassigned state.
+ */
+public class UnassignedMeta implements ToXContent {
+
+    public static final FormatDateTimeFormatter DATE_TIME_FORMATTER = Joda.forPattern("dateOptionalTime");
+
+    /**
+     * Reason why the shard is in unassigned state.
+     */
+    public enum Reason {
+        /**
+         * A fresh shard routing, never been assigned.
+         */
+        NEW((byte) 0),
```

Comment:
I wonder if it would be handy, for future changes, to make a distinction between shards that were just created via create index api calls (and thus can be assigned anywhere) and ones that were recovered from a cluster state on disk but still need assigned (for which we want to do some effort and find a copy)